File: tools/shell.ts
// shell — run a shell command (safe: false — blocked unless safeMode=false)
import type { Tool } from '../src/types.ts'

export const shell: Tool = {
  name: 'shell',
  description: 'Execute a shell command and return stdout/stderr',
  safe: false, // intentionally unsafe — enable by setting agent safeMode=false
  parameters: {
    type: 'object',
    properties: {
      command: { type: 'string', description: 'Shell command to run' },
    },
    required: ['command'],
  },
  async run({ command }) {
    const proc = Bun.spawn(['sh', '-c', command as string], {
      stdout: 'pipe',
      stderr: 'pipe',
    })
    const [out, err] = await Promise.all([
      new Response(proc.stdout).text(),
      new Response(proc.stderr).text(),
    ])
    const code = await proc.exited
    return [`exit=${code}`, out && `stdout:\n${out}`, err && `stderr:\n${err}`]
      .filter(Boolean)
      .join('\n')
      .trim()
  },
}

File: clients/python/marie/tools/shell.py
import subprocess
from .base import Tool

class ShellTool(Tool):
    def __init__(self):
        super().__init__(
            name="shell",
            description="Execute a shell command and return stdout/stderr. Use for system tasks.",
            parameters={
                "type": "object",
                "properties": {
                    "command": {"type": "string", "description": "The shell command to execute"}
                },
                "required": ["command"]
            },
            safe=False
        )

    def run(self, command):
        try:
            result = subprocess.run(
                command,
                shell=True,
                capture_output=True,
                text=True,
                timeout=30
            )
            output = f"exit_code: {result.returncode}\n"
            if result.stdout:
                output += f"stdout:\n{result.stdout}\n"
            if result.stderr:
                output += f"stderr:\n{result.stderr}\n"
            return output.strip()
        except subprocess.TimeoutExpired:
            return "Error: Command timed out after 30 seconds."
        except Exception as e:
            return f"Error executing shell: {str(e)}"
